cirq-core/cirq/sim/simulator.py

```diff
@@ -601,6 +673,67 @@ def _create_simulator_trial_result(
         """
         raise NotImplementedError()
 
+    def _run(
+        self, circuit: circuits.Circuit, param_resolver: study.ParamResolver, repetitions: int
+    ) -> Dict[str, np.ndarray]:
+        """See definition in `cirq.SimulatesSamples`."""
+        if self._ignore_measurement_results:
+            raise ValueError("run() is not supported when ignore_measurement_results = True")
+
+        param_resolver = param_resolver or study.ParamResolver({})
+        resolved_circuit = protocols.resolve_parameters(circuit, param_resolver)
+        check_all_resolved(resolved_circuit)
+        qubits = tuple(sorted(resolved_circuit.all_qubits()))
+        act_on_args = self._create_act_on_args(0, qubits)
+
+        prefix, general_suffix = (
+            split_into_matching_protocol_then_general(resolved_circuit, self._can_be_in_run_prefix)
+            if self._can_be_in_run_prefix(self.noise)
+            else (resolved_circuit[0:0], resolved_circuit)
+        )
+        step_result = None
+        for step_result in self._core_iterator(
+            circuit=prefix,
+            sim_state=act_on_args,
+        ):
+            pass
+
+        general_ops = list(general_suffix.all_operations())
+        if all(isinstance(op.gate, ops.MeasurementGate) for op in general_ops):
+            for step_result in self._core_iterator(
+                circuit=general_suffix,
+                sim_state=act_on_args,
+                all_measurements_are_terminal=True,
+            ):
+                pass
+            assert step_result is not None
+            measurement_ops = [cast(ops.GateOperation, op) for op in general_ops]
+            return step_result.sample_measurement_ops(measurement_ops, repetitions, seed=self._prng)
+        return self._run_sweep_repeat(general_suffix, repetitions, act_on_args)
+
+    def _run_sweep_repeat(
+        self,
+        circuit: circuits.Circuit,
+        repetitions: int,
+        act_on_args: TActOnArgs,
+    ) -> Dict[str, np.ndarray]:
+        measurements = {}  # type: Dict[str, List[np.ndarray]]
+
+        for _ in range(repetitions):
+            all_step_results = self._core_iterator(
+                circuit,
+                sim_state=act_on_args.copy(),
+            )
+            for step_result in all_step_results:
+                for k, v in step_result.measurements.items():
+                    if k not in measurements:
+                        measurements[k] = []
+                    measurements[k].append(np.array(v, dtype=np.uint8))
+        return {k: np.array(v) for k, v in measurements.items()}
+
+    def _can_be_in_run_prefix(self, val: Any):
```

Comment:
Please add a docstring for this method, since it also appears outside of this class.